### ISSUE TITLE
ScrollView.scroll_to broken

### ISSUE THREAD

**Issue by KeyWeeUsr** [kivy/kivy#4619]
```
from kivy.lang import Builder
from kivy.base import runTouchApp
from kivy.uix.boxlayout import BoxLayout
Builder.load_string('''
#:import Factory kivy.factory.Factory
<Test>:
    BoxLayout:
        ScrollView:
            id: scroll
            size_hint_x: 0.9
            GridLayout:
                id: grid
                rows: 1
                size_hint_x: None
                width: self.minimum_width
                Button:
                    size_hint_x: None
                    width: 50

        Button:
            text: '+'
            size_hint_x: 0.1
            on_release:
                grid.add_widget(Factory.Button(size_hint_x=None, width=50))
                scroll.scroll_to(grid.children[0])
''')
class Test(BoxLayout): pass
runTouchApp(Test())
```

This example work for vertical scrolling (if `width` -> `height` and `size_hint_x` -> `size_hint_y`), however it ignores the horizontal one. `scroll_x=1` is only a quick fix as it has no animation included.

**Comment by nadaz** [comment]
First thing, i wanted to say that how the example is written, seems the natural way to use scroll_to method.
But, i've looked a bit into it and this is not a scroll_to bug (_at least i think_). 
Two little example to explain:

1. If you put `print grid.children[0]` between grid.add_widget and scroll.scroll_to you'll see that it always print (0,0); this happen because the do_layout method is triggered before next frame (so you're calling scroll_to before the layout update).

2. Put the scroll_to line into schedule and you'll see that it works:
`Clock.schedule_once(lambda *dt:scroll.scroll_to(grid.children[0]),-1)`

So, i stopped here, because i could only think of forcing do_layout, and seemed excessive.
Anyway maybe someone can think a better solution.

**Comment by KeyWeeUsr** [comment]
@nadaz Makes sense as `(0, 0)` is actually at the bottom left, so it'll scroll always to the bottom left.

```
from kivy.lang import Builder
from kivy.base import runTouchApp
from kivy.uix.boxlayout import BoxLayout
Builder.load_string('''
#:import Factory kivy.factory.Factory
<Test>:
    BoxLayout:
        orientation: 'vertical'
        ScrollView:
            id: scroll
            size_hint_y: 0.9
            GridLayout:
                id: grid
                cols: 1
                size_hint_y: None
                height: self.minimum_height
                Button:
                    size_hint_y: None
                    height: 50

        Button:
            text: '+'
            size_hint_y: 0.1
            on_release:
                grid.add_widget(Factory.Button(text=str(len(grid.children)), size_hint_y=None, height=50), len(grid.children))
                print grid.children[-1].pos, grid.children[-1].text
                scroll.scroll_to(grid.children[-1])
''')
class Test(BoxLayout): pass
runTouchApp(Test())
```
I wonder, does it make the whole `scroll_to` method broken? The point is that the `ScrollView.scroll_to(<child widget>)` should work without using `Clock`. So a fix could be probably what you suggested - scheduled task with `Clock` before the next frame, however it should be part of the `ScrollView.scroll_to` method, not just a side note to use it with Clock.

**Comment by nadaz** [comment]
Well Clocking like you said and then deleted should do the trick (i haven't checked but i think schedule before next frame are executed in the order they are scheduled). I also tryed to put:
```
        if hasattr(self._viewport,'do_layout'):
            self._viewport.do_layout()
```
line 898 of scrollview, it's not "pretty" from my point of view, but it works. Test it if you want.
I would anyway ask to someone of the "big" developer if there's a cleaner way.

**Comment by KeyWeeUsr** [comment]
Yeah, you are right, the scheduling works. I made a typo, so the partial in `scroll_to` didn't call the `_scroll_to`(original method), which resulted in me thinking it doesn't work for some reason in `scrollview.py`, but in works in `main.py` or kv.

Anyway, the `hasattr` is a better trick and it's more visible what's actually happening than just creating probably even a race condition with Clock. This might be useful in the future is something similar comes up.

Now, could you make a Pull Request? It's your fix after all 😋 If you don't feel good [using git](https://kivy.org/docs/contribute.html) to contribute, just fork `kivy/kivy` and edit the file in your fork through github. Then go to `kivy/kivy` repo and a big green button saying something about creating pull request should show up. :)

**Comment by nadaz** [comment]
I would be happy to _really_ learn git, and contribute to kivy (as i can, obviously).
My concern about this particular case is that in my version, do_layout would be called twice (ok, only in the case of a kv-double instruction) and we have to think to large datasets, not small example; the clock method as you suggested would instead avoid this problem (sorry as i don't see the race condition, don't know Clock enough).
Just to be clear, i said that it would be better than the scheduling meaning it's easier to understand for someone getting into kivy source; but from a performance point of view, i'm not sure.
Anyway i'll give it a better look as soon as i can, (tomorrow) and then up the pull request if i don't see any better solution ;)

**Comment by KeyWeeUsr** [comment]
@nadaz I've tested the `do_layout` fix with cProfiler in a _superugly_ way (probably no one will ever use it like that) and the difference in performance was horrible (10-15min) 😃

Then again, we discussed the issue with @tshirtman and `Clock` is the way, yet... without `-1` as for "before the next frame". `Clock.schedule_once(lambda *dt:scroll.scroll_to(grid.children[0]))` which if you put it in scrollview.py will result in something similar. Don't forget to check if the clock calls the method like me (I forgot `*args` in the called one - the typo).
